File: app/src/main/java/bobrchess/of/by/belaruschess/presenter/impl/UserPresenterImpl.java
package bobrchess.of.by.belaruschess.presenter.impl;

import android.widget.TextView;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import bobrchess.of.by.belaruschess.R;
import bobrchess.of.by.belaruschess.dto.CountryDTO;
import bobrchess.of.by.belaruschess.dto.ErrorDTO;
import bobrchess.of.by.belaruschess.dto.RankDTO;
import bobrchess.of.by.belaruschess.dto.UserDTO;
import bobrchess.of.by.belaruschess.dto.extended.ExtendedUserDTO;
import bobrchess.of.by.belaruschess.exception.IncorrectDataException;
import bobrchess.of.by.belaruschess.network.connection.internal.SearchUserConnection;
import bobrchess.of.by.belaruschess.presenter.UserPresenter;
import bobrchess.of.by.belaruschess.presenter.callback.CallBackSearchUser;
import bobrchess.of.by.belaruschess.util.Util;
import bobrchess.of.by.belaruschess.util.Validator;
import bobrchess.of.by.belaruschess.view.activity.PackageModel;
import bobrchess.of.by.belaruschess.view.activity.UserContractView;
import butterknife.BindView;

import static bobrchess.of.by.belaruschess.util.Constants.TOKEN;

public class UserPresenterImpl implements CallBackSearchUser, UserPresenter {

    private UserContractView view;
    private SearchUserConnection userConnection;
    private Boolean viewIsReady = false;
    private PackageModel packageModel;
    private Integer connectivityStatus = 0;
    private Integer selectedRankIndex = 0;
    private Integer selectedCountryIndex = 0;
    private Integer selectedCoachIndex = 0;
    private Integer selectedGenderIndex = 0;
    private Map<Integer, RankDTO> ranksIndexes = new HashMap<>();
    private Map<Integer, CountryDTO> countriesIndexes = new HashMap<>();
    private Map<Integer, UserDTO> coachesIndexes = new HashMap<>();


    @BindView(R.id.t_link_registration)
    TextView registrationLink;

    public UserPresenterImpl() {
        userConnection = new SearchUserConnection();
        userConnection.attachPresenter(this);
    }

    @Override
    public void loadUserById(int id) {
        view.showProgress();
        userConnection.getUserById(packageModel.getValue(TOKEN), id);
    }

    @Override
    public void loadReferees() {
        view.showProgress();
        userConnection.getReferees(packageModel.getValue(TOKEN));
    }

    @Override
    public void loadCoaches() {
        view.showProgress();
        userConnection.getCoaches(packageModel.getValue(TOKEN));
    }

    @Override
    public void loadUsers() {
        view.showProgress();
        userConnection.getUsers(10);
    }//todo

    @Override
    public void loadUsers(Integer count) {
        view.showProgress();
        userConnection.getUsers(count);
    }

    @Override
    public void searchUsers(@NotNull String text) {
        if (viewIsReady) {
            view.showProgress();
            userConnection.searchUsers(text);
        }
    }

    @Override
    public void updateUser(@NotNull ExtendedUserDTO user) {
        if (viewIsReady) {
            //view.disableButton();
            try {
                user.setSelectedCoachIndex(selectedCoachIndex - 1);
                user.setSelectedCountryIndex(selectedCountryIndex);
                user.setSelectedRankIndex(selectedRankIndex);
                user.setSelectedGenderIndex(selectedGenderIndex);
                Validator.INSTANCE.validateUserData(user, false);
                user.setCoach(coachesIndexes.get(selectedCoachIndex - 1));
                user.setCountry(countriesIndexes.get(selectedCountryIndex));
                user.setRank(ranksIndexes.get(selectedRankIndex));
                user.setBeMale(selectedGenderIndex == 0);
                //user.setCoach(null);
                //   view.disableButton();
                view.showProgress();
                userConnection.updateUser(new UserDTO(user), packageModel.getValue(TOKEN));
            } catch (IncorrectDataException e) {
                view.showToast(e.getLocalizedMessage());//todo проверить что норм показывается и остальные так же плэйс и тд
                view.hideProgress();
            //    view.enableButton();
            }
        }
    }

    @Override
    public void attachView(@NotNull UserContractView userContractView) {
        view = userContractView;
    }

    @Override
    public void detachView() {
        view = null;
    }

    @Override
    public void viewIsReady() {
        viewIsReady = true;
    }

    @Override
    public void onResponse(@NotNull List<? extends UserDTO> users) {
        view.showUsers(users);
        view.hideProgress();
    }

    @Override
    public void onResponse(@NotNull UserDTO user) {
        view.showUser(user);
        view.hideProgress();
    }

    @Override
    public void onFailure(@NotNull ErrorDTO errorDTO) {
        view.hideProgress();
        view.showToast(errorDTO.getError());
    }

    @Override
    public void onServerUnavailable() {
        view.hideProgress();
    }

    @Override
    public void onUnsuccessfulRequest(@Nullable String message) {
        view.hideProgress();
    }

    @Override
    public void setConnectivityStatus(Integer status) {
        this.connectivityStatus = status;
    }

    public PackageModel getPackageModel() {
        return packageModel;
    }

    public void setPackageModel(PackageModel packageModel) {
        this.packageModel = packageModel;
    }

    @Override
    public boolean isConnected(int status) {
        return Util.Companion.isConnected(status);
    }

    @Override
    public void setSelectedRankIndex(int selectedRankIndex) {
        this.selectedRankIndex = selectedRankIndex;
    }

    @Override
    public void setSelectedCountryIndex(int selectedCountryIndex) {
        this.selectedCountryIndex = selectedCountryIndex;
    }

    @Override
    public void setSelectedCoachIndex(int selectedCoachIndex) {
        this.selectedCoachIndex = selectedCoachIndex;
    }

    @Override
    public void setSelectedGenderIndex(int selectedGenderIndex) {
        this.selectedGenderIndex = selectedGenderIndex;
    }

    @Override
    public void saveRanksIndexes(@Nullable List<RankDTO> ranks) {
        if (ranks != null) {
            for (int i = 0; i < ranks.size(); i++) {
                ranksIndexes.put(i, ranks.get(i));
            }
        }
    }

    @Override
    public void saveCoachesIndexes(@Nullable List<? extends UserDTO> coaches) {
        if (coaches != null) {
            for (int i = 0; i < coaches.size(); i++) {
                coachesIndexes.put(i, coaches.get(i));
            }
        }
    }

    @Override
    public void saveCountriesIndexes(List<CountryDTO> countries) {
        if (countries != null) {
            for (int i = 0; i < countries.size(); i++) {
                countriesIndexes.put(i, countries.get(i));
            }
        }
    }

}

File: app/src/main/java/bobrchess/of/by/belaruschess/presenter/impl/UserPresenterImpl.kt
package bobrchess.of.by.belaruschess.presenter.impl

import android.widget.TextView
import bobrchess.of.by.belaruschess.R
import bobrchess.of.by.belaruschess.dto.*
import bobrchess.of.by.belaruschess.dto.extended.ExtendedUserDTO
import bobrchess.of.by.belaruschess.exception.IncorrectDataException
import bobrchess.of.by.belaruschess.network.connection.internal.SearchUserConnection
import bobrchess.of.by.belaruschess.network.connection.internal.TokenConnection
import bobrchess.of.by.belaruschess.presenter.UserPresenter
import bobrchess.of.by.belaruschess.presenter.callback.CallBackSearchUser
import bobrchess.of.by.belaruschess.presenter.callback.CallBackToken
import bobrchess.of.by.belaruschess.util.Constants
import bobrchess.of.by.belaruschess.util.Util
import bobrchess.of.by.belaruschess.util.Validator.validateUserData
import bobrchess.of.by.belaruschess.view.activity.PackageModel
import bobrchess.of.by.belaruschess.view.activity.UserContractView
import butterknife.BindView
import org.springframework.util.StringUtils
import java.io.File

class UserPresenterImpl : CallBackSearchUser, CallBackToken, UserPresenter {
    private var view: UserContractView? = null
    private var userConnection: SearchUserConnection = SearchUserConnection()
    private var tokenConnection: TokenConnection = TokenConnection()
    private var viewIsReady = false
    private var packageModel: PackageModel? = null
    private var connectivityStatus: Int? = 0
    private var selectedRankIndex = 0
    private var selectedCountryIndex = 0
    private var selectedCoachIndex = 0
    private var selectedGenderIndex = 0
    private val ranksIndexes: MutableMap<Int, RankDTO?> = HashMap()
    private val countriesIndexes: MutableMap<Int, CountryDTO?> = HashMap()
    private val coachesIndexes: MutableMap<Int, String> = HashMap()
    var updatedUser: UserDTO? = null
    var updatedUserImage: File? = null

    @JvmField
    @BindView(R.id.t_link_registration)
    var registrationLink: TextView? = null

    override fun loadUserById(id: Int) {
        view!!.showProgress()
        userConnection.getUserById(packageModel!!.getSharePrefValue(Constants.TOKEN), id)
    }

    override fun loadReferees() {
        //view!!.showProgress()
        userConnection.getReferees(packageModel!!.getSharePrefValue(Constants.TOKEN))
    }

    override fun loadCoaches() {
      //  view!!.showProgress()
        userConnection.getCoaches()
    }

    override fun loadUsers(count: Int?) {
        view!!.showProgress()
        userConnection.getUsers(count, packageModel!!.getSharePrefValue(Constants.TOKEN))
    }

    override fun searchUsers(text: String) {
        if (viewIsReady) {
            view!!.showProgress()
            userConnection.searchUsers(text, packageModel!!.getSharePrefValue(Constants.TOKEN))
        }
    }

    override fun updateUser(user: ExtendedUserDTO, userImage: File?) {
        if (viewIsReady) {
            try {
                user.selectedCoachIndex = selectedCoachIndex - 1
                user.selectedCountryIndex = selectedCountryIndex
                user.selectedRankIndex = selectedRankIndex
                user.selectedGenderIndex = selectedGenderIndex
                validateUserData(user, false)
                user.coach = coachesIndexes[selectedCoachIndex - 1]
                user.country = countriesIndexes[selectedCountryIndex]
                user.rank = ranksIndexes[selectedRankIndex]
                user.beMale = selectedGenderIndex == 0
                view!!.showProgress()

                updatedUser = user
                updatedUserImage = Util.compressImage(userImage)

                userConnection.updateUser(updatedUser!!, updatedUserImage, packageModel!!.getSharePrefValue(Constants.TOKEN))
            } catch (e: IncorrectDataException) {
                view!!.showToast(e.localizedMessage)
                view!!.hideProgress()
            }
        }
    }

    override fun setPackageModel(packageModel: PackageModel?) {
        this.packageModel = packageModel
    }

    override fun attachView(userContractView: UserContractView) {
        view = userContractView
    }

    override fun detachView() {
        view = null
    }

    override fun viewIsReady() {
        viewIsReady = true
    }

    override fun onResponse(users: List<UserDTO>) {
        view!!.showUsers(users)
        view!!.hideProgress()
    }

    override fun onResponse(user: UserDTO) {
        view!!.showUser(user)
        view!!.hideProgress()
    }

    override fun onResponse(tokenDTO: TokenDTO) {
        packageModel!!.addSharePref(Constants.TOKEN, tokenDTO.token)

        if (updatedUser != null) {
            userConnection.updateUser(updatedUser!!, updatedUserImage, packageModel!!.getSharePrefValue(Constants.TOKEN))
        } else {
            loadReferees()
            loadCoaches()
        }
    }

    override fun onFailure(errorDTO: ErrorDTO) {
        view!!.hideProgress()
        when (errorDTO.error) {
            Constants.SERVER_UNAVAILABLE -> {
                onServerUnavailable()
            }
            Constants.KEY_UNSUCCESSFUL_REQUEST, Constants.INTERNAL_SERVER_ERROR -> {
                when (errorDTO.message) {
                    Constants.KEY_USER_ALREADY_EXISTS -> {
                        onUserExists()
                    }
                    else -> {
                        onUnsuccessfulRequest(Util.getInternalizedMessage(Constants.KEY_INTERNAL_SERVER_ERROR))
                    }
                }

            }
            else -> {
                when (errorDTO.message) {
                    Constants.TOKEN_IS_EXPIRED_MESSAGE -> {
                        view!!.showProgress()
                        refreshToken()
                    }
                    else -> {
                        onUnsuccessfulRequest(Util.getInternalizedMessage(Constants.KEY_INTERNAL_SERVER_ERROR))
                    }
                }
            }
        }
    }

    private fun refreshToken() {
        val refreshToken = packageModel!!.getSharePrefValue(Constants.REFRESH_TOKEN)
        if (!StringUtils.isEmpty(refreshToken)) {
            tokenConnection.refreshToken(refreshToken)
        } else {
            onUnsuccessfulRequest(Util.getInternalizedMessage(Constants.KEY_INTERNAL_SERVER_ERROR))
        }
    }

    private fun onUserExists() {
        view!!.showSnackbar(R.string.userAlreadyExists)
    }

    override fun onServerUnavailable() {
        view!!.hideProgress()
    }

    override fun onUnsuccessfulRequest(message: String?) {
        view!!.hideProgress()
    }

    override fun setConnectivityStatus(status: Int?) {
        connectivityStatus = status
    }

    override fun isConnected(status: Int): Boolean {
        return Util.isConnected(status)
    }

    override fun setSelectedRankIndex(selectedRankIndex: Int) {
        this.selectedRankIndex = selectedRankIndex
    }

    override fun setSelectedCountryIndex(selectedCountryIndex: Int) {
        this.selectedCountryIndex = selectedCountryIndex
    }

    override fun setSelectedCoachIndex(selectedCoachIndex: Int) {
        this.selectedCoachIndex = selectedCoachIndex
    }

    override fun setSelectedGenderIndex(selectedGenderIndex: Int) {
        this.selectedGenderIndex = selectedGenderIndex
    }

    override fun saveRanksIndexes(ranks: List<RankDTO?>?) {
        if (ranks != null) {
            for (i in ranks.indices) {
                ranksIndexes[i] = ranks[i]
            }
        }
    }

    override fun saveCoachesIndexes(coaches: List<UserDTO?>?) {
        if (coaches != null) {
            for (i in coaches.indices) {
                var rank = coaches[i]!!.rank?.abbreviation
                rank = if (!StringUtils.isEmpty(rank)) " ($rank)" else ""
                coachesIndexes[i] = coaches[i]!!.name + " " + coaches[i]!!.surname + rank
            }
        }
    }

    override fun saveCountriesIndexes(countries: List<CountryDTO?>?) {
        if (countries != null) {
            for (i in countries.indices) {
                countriesIndexes[i] = countries[i]
            }
        }
    }

    init {
        userConnection.attachPresenter(this)
        tokenConnection.attachPresenter(this)
    }
}